### PR TITLE
SIP 279 Slippage Protection

### PR DESCRIPTION
**Pull request by davidvuong**
As part of SIP 279, we introduce slippage protection as a complementary mechanism to premium/discount pricing. Slippage protection will allow users to specify a price percentage delta when making trades in both delayed orders and regular orders. The most impactful change involves a modification to two public interface methods modifyPosition and submitDelayedOrder.

```solidity
interface PerpsV2Market {
	function modifyPosition(int sizeDelta, uint slippage);
	function submitDelayedOrder(int sizeDelta, uint slippage, uint desiredTimeDelta);
        function submitDelayedOrderWithTracking(int sizeDelta, uint slippage, uint desiredTimeDelta, bytes32 trackingCode);
        function submitOffchainDelayedOrder(int sizeDelta, uint slippage);
        function submitOffchainDelayedOrderWithTracking(
            int sizeDelta,
            uint slippage,
            bytes32 trackingCode
        );
}
```

Example:

```
ETH Price (price)  = 1000
Slippage           = 0.01 (1%)
p/d Price (price') = 1005

Slippage Upper Bound = 1000 * (1 + 0.01)
                     = 1010
```

If price' is under the slippage upper bound then the trade can proceed to execute. If not, the execution is reverted.

Moreover, when a `0` slippage is provided, we default the slippage price to be `currentPrice + maker/takerFee` instead of `currentPrice + (1 + slippage)`.

Note: Unit tests should all be passing. However, I have yet to implement new unit tests to account for this new feature. That will come in a later PR.

This also updates the `postTradeDetails` function to allow a user to specify an arbitrary `tradePrice`.

### CODE REVIEW

File: test/contracts/PerpsV2Market.offchainDelayedOrder.js [hunk 1]
```diff
@@ -37,6 +37,7 @@ contract('PerpsV2Market PerpsV2MarketOffchainOrders', accounts => {
 	const takerFeeOffchainDelayedOrder = toUnit('0.00005');
 	const makerFeeOffchainDelayedOrder = toUnit('0.00001');
 	const initialPrice = toUnit('100');
+	const slippage = toUnit('0.5'); // 50% slippage tolerance acceptable.
```
Comment (DariusPrzydzial) [inline test/contracts/PerpsV2Market.offchainDelayedOrder.js:40]: 50% or 50bps?

**Review by DariusPrzydzial — COMMENTED** [review]
read diffs, one comment

**Review by leomassazza — APPROVED** [review]
LGTM

File: test/contracts/PerpsV2Market.offchainDelayedOrder.js (hunk 1 above)
Comment (davidvuong) [reply test/contracts/PerpsV2Market.offchainDelayedOrder.js]: 500bps. I've updated comments to be clearer.

A 50bps would be 0.5% as a percentage and 0.005 as a BN unit here. It's the same units as how maker/taker fees are defined currently.

**Pull request merged**